test/e2e/utils/cnpspecbuilder.go

```diff
@@ -0,0 +1,214 @@
+	podSelectorMatchExp *[]metav1.LabelSelectorRequirement, nsSelectorMatchExp *[]metav1.LabelSelectorRequirement,
+	action secv1alpha1.RuleAction) *ClusterNetworkPolicySpecBuilder {
+
+	var ps *metav1.LabelSelector
+	var ns *metav1.LabelSelector
+	if b.Spec.Ingress == nil {
+		b.Spec.Ingress = []secv1alpha1.Rule{}
+	}
+
+	if podSelector != nil {
+		ps = &metav1.LabelSelector{
+			MatchLabels: podSelector,
+		}
+		if podSelectorMatchExp != nil {
+			ps.MatchExpressions = *podSelectorMatchExp
+		}
+	}
+	if podSelectorMatchExp != nil {
+		ps = &metav1.LabelSelector{
+			MatchExpressions: *podSelectorMatchExp,
+		}
+	}
+	if nsSelector != nil {
+		ns = &metav1.LabelSelector{
+			MatchLabels: nsSelector,
+		}
+		if nsSelectorMatchExp != nil {
+			ns.MatchExpressions = *nsSelectorMatchExp
+		}
+	}
+	if nsSelectorMatchExp != nil {
+		ns = &metav1.LabelSelector{
+			MatchExpressions: *nsSelectorMatchExp,
+		}
+	}
+	var ipBlock *secv1alpha1.IPBlock
+	if cidr != nil {
+		ipBlock = &secv1alpha1.IPBlock{
+			CIDR: *cidr,
+		}
+	}
+	var policyPeer []secv1alpha1.NetworkPolicyPeer
+	if ps != nil || ns != nil || ipBlock != nil {
+		policyPeer = []secv1alpha1.NetworkPolicyPeer{{
+			PodSelector:       ps,
+			NamespaceSelector: ns,
+			IPBlock:           ipBlock,
+		}}
+	}
+
+	var ports []secv1alpha1.NetworkPolicyPort
+	if port != nil && portName != nil {
+		panic("specify portname or port, not both")
+	}
+	if port != nil {
+		ports = []secv1alpha1.NetworkPolicyPort{
+			{
+				Port:     &intstr.IntOrString{IntVal: int32(*port)},
+				Protocol: &protoc,
+			},
+		}
+	}
+	if portName != nil {
+		ports = []secv1alpha1.NetworkPolicyPort{
+			{
+				Port:     &intstr.IntOrString{Type: intstr.String, StrVal: *portName},
+				Protocol: &protoc,
+			},
+		}
+	}
+	newRule := secv1alpha1.Rule{
+		From:   policyPeer,
+		Ports:  ports,
+		Action: &action,
+	}
+	b.Spec.Ingress = append(b.Spec.Ingress, newRule)
+	return b
+}
+
+func (b *ClusterNetworkPolicySpecBuilder) AddEgress(protoc v1.Protocol,
+	port *int, portName *string, cidr *string,
+	podSelector map[string]string, nsSelector map[string]string,
+	podSelectorMatchExp *[]metav1.LabelSelectorRequirement, nsSelectorMatchExp *[]metav1.LabelSelectorRequirement,
+	action secv1alpha1.RuleAction) *ClusterNetworkPolicySpecBuilder {
+
+	// For simplicity, we just reuse the Ingress code here.  The underlying data model for ingress/egress is identical
+	// With the exception of calling the rule `To` vs. `From`.
+	c := &ClusterNetworkPolicySpecBuilder{}
+	c.AddIngress(protoc, port, portName, cidr, podSelector, nsSelector, podSelectorMatchExp, nsSelectorMatchExp, action)
+	theRule := c.Get().Spec.Ingress[0]
+
+	b.Spec.Egress = append(b.Spec.Egress, secv1alpha1.Rule{
+		To:     theRule.From,
+		Ports:  theRule.Ports,
+		Action: theRule.Action,
+	})
+	return b
+}
+
+// AddEgressDNS mutates the nth policy rule to allow DNS, convenience method
+func (b *ClusterNetworkPolicySpecBuilder) WithEgressDNS() *ClusterNetworkPolicySpecBuilder {
```

Comment:
name and behavior doesn't match comment, it doesn't mutates the nth rule only, and there is no place to provide n.